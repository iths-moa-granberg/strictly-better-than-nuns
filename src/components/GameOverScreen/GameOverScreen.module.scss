@import '../../scss/utils';

.gameOverWrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100vw;
  height: 100vh;
  color: $white;
}

.gameOverHeading {
  margin: 0;
  font-size: 5rem;
}

.gameOverWinner {
  margin: 0 0 2rem 0;
  font-size: 2rem;
}

.player-e1 {
  color: $enemy1-main-color;
  font-size: 3rem;
}

@for $i from 1 through length($playerColors) {
  .player-#{$i} {
    color: rgba(nth($playerColors, $i), 1);
    font-size: 3rem;
  }
}
