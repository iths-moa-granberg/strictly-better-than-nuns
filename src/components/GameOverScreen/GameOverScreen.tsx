import React from 'react';

import { ClientUser } from '../../shared/sharedTypes';

import styles from './GameOverScreen.module.scss';
import buttonStyles from '../../scss/Buttons.module.scss';

interface GameOverScreenProps {
  readonly winners: ClientUser[];
}

const formatWinners = (winners: ClientUser[]) => {
  const nameHtml = [];

  for (let i = 0; i < winners.length; i++) {
    const winner = winners[i];

    if (i >= 1) {
      if (i === winners.length - 1) {
        nameHtml.push(' and ');
      } else {
        nameHtml.push(', ');
      }
    }

    nameHtml.push(
      <span key={winner.userID} className={styles[`player-${winner.userID}`]}>
        {winner.username}
      </span>
    );
  }

  return nameHtml;
};

const GameOverScreen = ({ winners }: GameOverScreenProps) => {
  const handleClick = () => {
    window.location.reload();
  };

  return (
    <div className={styles.gameOverWrapper}>
      <h1 className={styles.gameOverHeading}>Game over!</h1>
      <h2 className={styles.gameOverWinner}>
        The {winners.length === 1 ? 'winner is' : 'winners are'} {formatWinners(winners)}!
      </h2>
      <button className={`${buttonStyles.button} ${buttonStyles.big}`} onClick={handleClick}>
        New game
      </button>
    </div>
  );
};

export default GameOverScreen;
