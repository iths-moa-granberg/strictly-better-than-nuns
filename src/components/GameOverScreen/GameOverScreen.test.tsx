import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import GameOverScreen from './GameOverScreen';

describe('GameOverScreen component', () => {
  it('renders with one winner', () => {
    const winners = [{ username: 'winner', userID: 'a1' }];
    const { container } = render(<GameOverScreen winners={winners} />);

    expect(container).toMatchSnapshot();
  });

  it('renders with two winners', () => {
    const winners = [
      { username: 'winner', userID: 'a1' },
      { username: 'winner nr 2', userID: 'b2' },
    ];
    const { container } = render(<GameOverScreen winners={winners} />);

    expect(container).toMatchSnapshot();
  });

  it('renders with multiple winners', () => {
    const winners = [
      { username: 'winner', userID: 'a1' },
      { username: 'winner nr 2', userID: 'b2' },
      { username: 'winner nr 3', userID: 'c3' },
      { username: 'winner nr 4', userID: 'd4' },
    ];
    const { container } = render(<GameOverScreen winners={winners} />);

    expect(container).toMatchSnapshot();
  });

  it('should reload on click', () => {
    const winners = [{ username: 'winner', userID: 'a1' }];
    const { getByText } = render(<GameOverScreen winners={winners} />);
    window.location.reload = jest.fn();

    fireEvent.click(getByText('New game'));

    expect(window.location.reload).toHaveBeenCalled();
  });
});
