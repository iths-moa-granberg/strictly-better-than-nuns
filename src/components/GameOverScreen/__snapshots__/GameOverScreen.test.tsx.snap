// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOverScreen component renders with multiple winners 1`] = `
<div>
  <div
    class="gameOverWrapper"
  >
    <h1
      class="gameOverHeading"
    >
      Game over!
    </h1>
    <h2
      class="gameOverWinner"
    >
      The 
      winners are
       
      <span
        class="player-a1"
      >
        winner
      </span>
      , 
      <span
        class="player-b2"
      >
        winner nr 2
      </span>
      , 
      <span
        class="player-c3"
      >
        winner nr 3
      </span>
       and 
      <span
        class="player-d4"
      >
        winner nr 4
      </span>
      !
    </h2>
    <button
      class="button big"
    >
      New game
    </button>
  </div>
</div>
`;

exports[`GameOverScreen component renders with one winner 1`] = `
<div>
  <div
    class="gameOverWrapper"
  >
    <h1
      class="gameOverHeading"
    >
      Game over!
    </h1>
    <h2
      class="gameOverWinner"
    >
      The 
      winner is
       
      <span
        class="player-a1"
      >
        winner
      </span>
      !
    </h2>
    <button
      class="button big"
    >
      New game
    </button>
  </div>
</div>
`;

exports[`GameOverScreen component renders with two winners 1`] = `
<div>
  <div
    class="gameOverWrapper"
  >
    <h1
      class="gameOverHeading"
    >
      Game over!
    </h1>
    <h2
      class="gameOverWinner"
    >
      The 
      winners are
       
      <span
        class="player-a1"
      >
        winner
      </span>
       and 
      <span
        class="player-b2"
      >
        winner nr 2
      </span>
      !
    </h2>
    <button
      class="button big"
    >
      New game
    </button>
  </div>
</div>
`;
